### PR TITLE
ssa: implement `#[collapse_debuginfo]`

### PR DESCRIPTION
**Pull request by davidtwco**
cc #39153 rust-lang/compiler-team#386

Debuginfo line information for macro invocations are collapsed by default - line information are replaced by the line of the outermost expansion site. Using `-Zdebug-macros` disables this behaviour.

When the `collapse_debuginfo` feature is enabled, the default behaviour is reversed so that debuginfo is not collapsed by default. In addition, the `#[collapse_debuginfo]` attribute is available and can be applied to macro definitions which will then have their line information collapsed.

r? rust-lang/wg-debugging

### CODE REVIEW

**Comment by davidtwco** [comment]
Relevant Zulip discussions are [here](https://rust-lang.zulipchat.com/#narrow/stream/233931-t-compiler.2Fmajor-changes/topic/Debuggable.20Macro.20Expansions.20compiler-team.23386) and [here](https://rust-lang.zulipchat.com/#narrow/stream/317568-t-compiler.2Fwg-debugging/topic/.5Btriage.20mtg.5D.202022-07-18/near/289969262).

**Comment by davidtwco** [comment]
It might make sense to add the `#[collapse_debuginfo]` annotations to std macros in this pull request, happy to do that if people agree.

File: compiler/rustc_expand/src/base.rs [hunk 1]
```diff
@@ -697,6 +697,9 @@ pub struct SyntaxExtension {
     pub allow_internal_unsafe: bool,
     /// Enables the macro helper hack (`ident!(...)` -> `$crate::ident!(...)`) for this macro.
     pub local_inner_macros: bool,
+    /// Should debuginfo for the macro be collapsed to the outermost expansion site (in other
+    /// words, was the macro definition annotated with `#[collapse_debuginfo]`)?
+    pub collapse_debuginfo: bool,
```
Comment (bjorn3) [inline compiler/rustc_expand/src/base.rs:702]: Should we use bitflags for `allow_internal_unsafe`, `local_inner_macros` and `collapse_debuginfo`?

File: compiler/rustc_codegen_ssa/src/mir/debuginfo.rs [hunk 2]
```diff
@@ -73,35 +73,36 @@ impl<'tcx, S: Copy, L: Copy> DebugScope<S, L> {
...
+    /// locations of macro expansions with that of the outermost expansion site (when the macro is
+    /// annotated with `#[collapse_debuginfo]` or when `-Zdebug-macros` is provided).
+    fn adjust_span_for_debugging(&self, bx: &Bx, mut span: Span) -> Span {
         // Bail out if debug info emission is not enabled.
         if self.debug_context.is_none() {
             return span;
         }
 
-        if span.from_expansion() && !self.cx.sess().opts.unstable_opts.debug_macros {
+        if bx.tcx().should_collapse_debuginfo(span) {
```
[the remark this answers is not in the document]
Comment (davidtwco) [reply compiler/rustc_codegen_ssa/src/mir/debuginfo.rs:105]: [`FunctionCx`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_codegen_ssa/mir/struct.FunctionCx.html) doesn't have a `tcx` member?

[the remark this answers is not in the document]
Comment (bjorn3) [reply compiler/rustc_codegen_ssa/src/mir/debuginfo.rs:105]: Try `self.cx.tcx()` then.

[the remark this answers is not in the document]
Comment (davidtwco) [reply compiler/rustc_codegen_ssa/src/mir/debuginfo.rs]: Hadn't spotted this possibility, fixed.

File: compiler/rustc_middle/src/ty/mod.rs [hunk 3]
```diff
@@ -2254,6 +2254,23 @@ impl<'tcx> TyCtxt<'tcx> {
...
+    ///
+    /// - If the `collapse_debuginfo` feature is enabled then debuginfo is not collapsed by default
+    ///   and only when a macro definition is annotated with `#[collapse_debuginfo]`.
+    /// - If `collapse_debuginfo` is not enabled, then debuginfo is collapsed by default.
+    ///
+    /// When `-Zdebug-macros` is provided then debuginfo will never be collapsed.
+    pub fn should_collapse_debuginfo(self, span: Span) -> bool {
+        if self.features().collapse_debuginfo {
+            span.in_macro_expansion_with_collapse_debuginfo()
+                && !self.sess.opts.unstable_opts.debug_macros
```
Comment (bjorn3) [inline compiler/rustc_middle/src/ty/mod.rs:2268]: Maybe check `self.sess.opts.unstable_opts.debug_macros` first? That should be a bit cheaper.

File: compiler/rustc_expand/src/base.rs (hunk 1 above)
Comment (davidtwco) [reply compiler/rustc_expand/src/base.rs]: Switched to using bitflags for this - decided to share the bitflags between `SyntaxExtension` and `ExpnData`, so ended up undoing the changes to `ExpnKind` that I had.

File: compiler/rustc_middle/src/ty/mod.rs (hunk 3 above)
Comment (davidtwco) [reply compiler/rustc_middle/src/ty/mod.rs]: Fixed!

**Comment by michaelwoerister** [comment]
Thanks for the PR, @davidtwco! I've set aside some time to read up on the discussion early next week. I guess, technically, we'll need to wait for the MCP to be accepted anyway `:)`

**Comment by michaelwoerister** [comment]
Alright, I'll finally have some time to take a look at this topic! Sorry for the wait, @davidtwco.

File: compiler/rustc_feature/src/active.rs [hunk 4]
```diff
@@ -336,6 +336,8 @@ declare_features! (
     (active, closure_track_caller, "1.57.0", Some(87417), None),
     /// Allows to use the `#[cmse_nonsecure_entry]` attribute.
     (active, cmse_nonsecure_entry, "1.48.0", Some(75835), None),
+    /// Allows use of the `#[collapse_debuginfo]` attribute.
+    (active, collapse_debuginfo, "1.62.0", None, None),
```
Comment (michaelwoerister) [inline compiler/rustc_feature/src/active.rs:340]: The version number should be current nightly, I think. 
See https://rustc-dev-guide.rust-lang.org/feature-gates.html#adding-a-feature-gate

File: compiler/rustc_passes/src/check_attr.rs [hunk 5]
```diff
@@ -130,6 +130,7 @@ impl CheckAttrVisitor<'_> {
                 | sym::rustc_if_this_changed
                 | sym::rustc_then_this_would_need => self.check_rustc_dirty_clean(&attr),
                 sym::cmse_nonsecure_entry => self.check_cmse_nonsecure_entry(attr, span, target),
+                sym::collapse_debuginfo => self.check_collapse_debuginfo(attr, span, target),
                 sym::const_trait => self.check_const_trait(attr, span, target),
                 sym::must_not_suspend => self.check_must_not_suspend(&attr, span, target),
                 sym::must_use => self.check_must_use(hir_id, &attr, span, target),
```
Comment (michaelwoerister) [inline compiler/rustc_passes/src/check_attr.rs:136]: This linear search through attributes annoys me every time I come across it 😅

File: src/test/debuginfo/collapse-debuginfo-no-attr-flag.rs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    () => {
+        two(); // #loc2
+    };
+}
+
+fn main() {
+    let ret = 0; // #break
+    outer!({
+        three(); // #loc3
+        four(); // #loc4
```
Comment (michaelwoerister) [inline src/test/debuginfo/collapse-debuginfo-no-attr-flag.rs:58]: I didn't know compiletest could do that. Neat!

File: src/test/debuginfo/collapse-debuginfo-with-attr-flag.rs [hunk 7]
```diff
@@ -0,0 +1,63 @@
+// ignore-lldb
+#![feature(collapse_debuginfo)]
+
+// Test that line numbers are not replaced with those of the outermost expansion site when the
+// `collapse_debuginfo` is active and `-Zdebug-macros` is provided, despite `#[collapse_debuginfo]`
+// being used.
```
Comment (michaelwoerister) [inline src/test/debuginfo/collapse-debuginfo-with-attr-flag.rs:6]: This is exactly the same behavior as `src/test/debuginfo/collapse-debuginfo-no-attr.rs`, right? Maybe use test revisions, so the file is not duplicated.

**Review by michaelwoerister — COMMENTED** [review]
This looks excellent! Great tests. Thanks for the PR, @davidtwco!

I left a few comments below. Looks like a tracking issue is needed for CI to pass.

**Comment by michaelwoerister** [comment]
From the MCP it looks like we want this to eventually be the default behavior, right? I.e. we'd annotate `assert!()`, `println!()`, etc with `#[collapse_debuginfo]` and the attribute would be available to end-users too, right?

**Comment by davidtwco** [comment]
> From the MCP it looks like we want this to eventually be the default behavior, right? I.e. we'd annotate `assert!()`, `println!()`, etc with `#[collapse_debuginfo]` and the attribute would be available to end-users too, right?

Yeah, the idea is that this feature would eventually be stabilized, and in doing so, that would change the default behaviour and make the attribute available to users. It would need to be added to the std library before that though.

File: compiler/rustc_feature/src/active.rs (hunk 4 above)
Comment (davidtwco) [reply compiler/rustc_feature/src/active.rs]: Updated, it was just outdated because of when the PR was opened..

**Comment by davidtwco** [comment]
> Looks like a tracking issue is needed for CI to pass.

Created #100758 and added it to the code so CI can continue.

File: src/test/debuginfo/collapse-debuginfo-with-attr-flag.rs (hunk 7 above)
Comment (davidtwco) [reply src/test/debuginfo/collapse-debuginfo-with-attr-flag.rs:6]: I think it's clearer if the tests are kept separate, while the behaviour is the same, the code itself isn't - this test uses the `#[collapse_debuginfo]` attribute while the `no-attr` test does not.

File: compiler/rustc_span/src/hygiene.rs [hunk 8]
```diff
@@ -903,6 +903,21 @@ impl Span {
     }
 }
 
+bitflags::bitflags! {
+    /// Flags for the syntax extension/expansion (e.g. should debuginfo be collapsed? should the
+    /// `unsafe_code` lint be suppressed?).
+    #[derive(Encodable, Decodable, HashStable_Generic)]
+    pub struct ExpnFlags: u32 {
```
Comment (klensy) [inline compiler/rustc_span/src/hygiene.rs:910]: One byte should be enough?
```suggestion
    pub struct ExpnFlags: u8 {
```

Comment (davidtwco) [reply compiler/rustc_span/src/hygiene.rs]: Yeah, I think so, fixed.

**Comment by michaelwoerister** [comment]
I'd guess that stabilizing the attribute would need to be approve by T-lang, right?

**Comment by davidtwco** [comment]
> I'd guess that stabilizing the attribute would need to be approve by T-lang, right?

That's my understanding from https://github.com/rust-lang/rfcs/pull/2117#issuecomment-718210420

File: src/test/debuginfo/collapse-debuginfo-with-attr-flag.rs (hunk 7 above)
Comment (michaelwoerister) [reply src/test/debuginfo/collapse-debuginfo-with-attr-flag.rs:6]: OK, that works for me too.

**Comment by michaelwoerister** [comment]
OK, let's do a perf run so we don't get any surprises. Other than that, you can consider the PR r=me.

Thanks a lot for pushing this over the finish line, @davidtwco!

@bors try @rust-timer queue

**Comment by davidtwco** [comment]
Just a conflict with another change, fixed now.

@bors try @rust-timer queue

**Review by wesleywiser — APPROVED** [review]

**Comment by wesleywiser** [comment]
Thanks @davidtwco! 👍 

@bors r+

**Pull request merged**